xla/service/float8_fnuz_ir_emitter.cc

```diff
@@ -0,0 +1,646 @@
+}
+
+// Returns the exponent of the input value shifted down to the least
+// significant bits and without any bias.
+llvm::Value* ExtractExponent(PrimitiveType type, llvm::Value* value,
+                             llvm::IRBuilder<>* b) {
+  const uint64_t shift = BitWidth(type) - 1;
+  const uint64_t bit_mask = (0x1ull << shift) - 0x1ull;
+  llvm::Type* uint_type = b->getIntNTy(BitWidth(type));
+
+  // Mask out sign bit.
+  llvm::Value* exponent = b->CreateAnd(value, bit_mask);
+
+  // Shift the mantissa bits away, leaving the exponent.
+  exponent = b->CreateLShr(
+      exponent, llvm::ConstantInt::get(uint_type, SignificandWidth(type) - 1));
+
+  // Subtract the exponent bias.
+  exponent = b->CreateSub(
+      exponent,
+      llvm::ConstantInt::get(uint_type, primitive_util::ExponentBias(type)));
+
+  // If the input number is not a normal number, return the subnormal exponent.
+  llvm::Value* input_normal_pred = IsNormalNumber(type, value, b);
+  return b->CreateSelect(
+      input_normal_pred, exponent,
+      llvm::ConstantInt::get(uint_type, UnderflowExponent(type) - 1));
+}
+
+// Returns the mantissa of the input value with all bits explicitly
+// represented. For normal numbers, the implicit leading 1 is in the
+// returned value.
+llvm::Value* ExtractMantissa(PrimitiveType type, llvm::Value* value,
+                             llvm::IRBuilder<>* b) {
+  const uint64_t shift = SignificandWidth(type) - 1;
+  const uint64_t mantissa_bit_mask = (0x1ull << shift) - 0x1ull;
+
+  llvm::Value* mantissa = b->CreateAnd(value, mantissa_bit_mask);
+
+  llvm::Value* input_normal_pred = IsNormalNumber(type, value, b);
+  llvm::Value* mantissa_normal = b->CreateOr(mantissa, (0x1ull << shift));
+
+  return b->CreateSelect(input_normal_pred, mantissa_normal, mantissa);
+}
+
+// Identifies the index of the last bit of the input that can be represented
+// in the output type. The index starts from the least significant bit
+// (index 0) to the most significant bit (bit n-1). This takes into account
+// whether the input value is a normal number.
+//
+// Example 1:
+// input_type  = F16
+// output_type = F8E5M2FNUZ
+// value = 0.00002664
+//       = 0x1.BFp-16
+//       = 0x01BF
+//       = 0b0|00000|0110111111
+//       = 0b0.0110111111 * 2^(-14)
+// Given the input and output is a denorm, we are looking for the bit the
+// corresponds to the smallest non-zero value. For F8E5M2FNUZ that's 2^(-17).
+// ExtractMantissa(value) = 0b0000000110111111
+//                                    ^- 2^(-17) is here at bit 7.
+// result = LastMantissaBit(F16, 0.00002664, F8E5M2FNUZ, b) = 7
+//
+// Example 2:
+// input_type  = BF16
+// output_type = F8E4M3FNUZ
+// value = 247.0
+//       = 0x1.EEp7
+//       = 0x4377
+//       = 0b0|10000110|1110111
+//       = 0b1.1110111 * 2^7
+// Given the input and output is a normal number, we are looking for the bit
+// the corresponds to the third bit of the mantissa.
+// ExtractMantissa(value) = 0b0000000011110111
```

Comment:
This value doesn't correspond to `0b0|10000110|1110111` above.

Reply:
Keep in mind, for normal numbers `ExtractMantissa` includes the leading `1`.